server/src/main/java/org/spine3/server/entity/AbstractEntity.java

```diff
@@ -0,0 +1,277 @@
+        this.version = version;
+        this.whenModified = checkNotNull(whenLastModified);
+    }
+
+    /**
+     * Updates the state incrementing the version number and recording time of the modification.
+     *
+     * @param newState a new state to set
+     */
+    protected void incrementState(S newState) {
+        setState(newState, incrementVersion(), getCurrentTime());
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public int getVersion() {
+        return version;
+    }
+
+    /**
+     * Advances the current version by one and records the time of the modification.
+     *
+     * @return new version number
+     */
+    protected int incrementVersion() {
+        ++version;
+        whenModified = getCurrentTime();
+        return version;
+    }
+
+    /**
+     * {@inheritDoc}
+     * @see #setState(Message, int, Timestamp)
+     */
+    @Override
+    @CheckReturnValue
+    public Timestamp whenModified() {
+        final Timestamp result = whenModified == null
+                                 ? Timestamp.getDefaultInstance()
+                                 : whenModified;
+        return result;
+    }
+
+    /**
+     * Obtains the entity status.
+     */
+    protected EntityStatus getStatus() {
+        final EntityStatus result = this.status ==  null
+                ? EntityStatus.getDefaultInstance()
+                : this.status;
+        return result;
+    }
+
+    /**
+     * Tests whether the entity is marked as archived.
+     *
+     * @return {@code true} if the entity is archived, {@code false} otherwise
+     */
+    protected boolean isArchived() {
+        return getStatus().getArchived();
+    }
+
+    /**
+     * Sets {@code archived} status flag to the passed value.
+     */
+    protected void setArchived(boolean archived) {
+        this.status = getStatus().toBuilder()
+                                 .setArchived(archived)
+                                 .build();
+    }
+
+    /**
+     * Tests whether the entity is marked as deleted.
+     *
+     * @return {@code true} if the entity is deleted, {@code false} otherwise
+     */
+    protected boolean isDeleted() {
+        return getStatus().getDeleted();
+    }
+
+    /**
+     * Sets {@code deleted} status flag to the passed value.
+     */
+    protected void setDeleted(boolean deleted) {
+        this.status = getStatus().toBuilder()
+                                 .setDeleted(deleted)
+                                 .build();
+    }
+
+    /**
+     * Ensures that the entity is not marked as {@code archived}.
+     *
+     * @throws CannotModifyArchivedEntity if the entity in in the archived status
+     * @see #getStatus()
+     * @see EntityStatus#getArchived()
+     */
+    protected void checkNotArchived() throws CannotModifyArchivedEntity {
+        if (getStatus().getArchived()) {
+            final String idStr = Stringifiers.idToString(getId());
+            throw new CannotModifyArchivedEntity(idStr);
```

Comment:
These methods were added as convenience syntax for aggregates that check their status while accepting modifications commands. I guess, they would move to a utility class preserving the names. I'd rather have more specific exceptions than one, which is too general. It would be better for diagnostics.